Storage/Rust/RustLogins.swift

```diff
@@ -815,75 +827,106 @@ public class RustLogins {
                                category: .storage)
                     GleanMetrics.LoginsStoreKeyRegeneration.corrupt.record()
-                    _ = self.wipeLocalEngine()
-
-                    return try rustKeys.createAndStoreKey()
+                    self.resetLoginsAndKey(rustKeys: rustKeys, completion: completion)
                 }
             } catch let error as NSError {
-                logger.log("Error retrieving logins encryption key",
+                logger.log("Error validating logins encryption key",
                            level: .warning,
                            category: .storage,
                            description: error.localizedDescription)
+                completion(.failure(error))
             }
         case (.some(key), .none):
             // The key is present, but we didn't expect it to be there.
-            do {
-                logger.log("Logins key lost due to storage malfunction, new one generated",
-                           level: .warning,
-                           category: .storage)
-                GleanMetrics.LoginsStoreKeyRegeneration.other.record()
-                _ = self.wipeLocalEngine()
 
-                return try rustKeys.createAndStoreKey()
-            } catch let error as NSError {
-                throw error
-            }
+            logger.log("Logins key lost due to storage malfunction, new one generated",
+                       level: .warning,
+                       category: .storage)
+            GleanMetrics.LoginsStoreKeyRegeneration.other.record()
+            self.resetLoginsAndKey(rustKeys: rustKeys, completion: completion)
         case (.none, .some(encryptedCanaryPhrase)):
             // We expected the key to be present, but it's gone missing on us.
-            do {
-                logger.log("Logins key lost, new one generated",
-                           level: .warning,
-                           category: .storage)
-                GleanMetrics.LoginsStoreKeyRegeneration.lost.record()
-                _ = self.wipeLocalEngine()
 
-                return try rustKeys.createAndStoreKey()
-            } catch let error as NSError {
-                throw error
-            }
+            logger.log("Logins key lost, new one generated",
+                       level: .warning,
+                       category: .storage)
+            GleanMetrics.LoginsStoreKeyRegeneration.lost.record()
+            self.resetLoginsAndKey(rustKeys: rustKeys, completion: completion)
         case (.none, .none):
-            // We didn't expect the key to be present, and it's not (which is the case for first-time calls).
-            do {
-                return try rustKeys.createAndStoreKey()
-            } catch let error as NSError {
-                throw error
+            // We didn't expect the key to be present, which either means this is a first-time
+            // call or the key data has been cleared from the keychain.
```

Comment:
Is it possible that the encryption data is unavailable instead of missing in this instance? Is there an availability check I can do for the keychain?

Reply:
I've worked with keychain very little so I can't answer this question without making stuff up. Lemme know if you want me to look into Apple's docs to find out more for you.

Reply:
Yeah we probably need an answer to this question because if the keychain is unavailable we're wiping data without cause. It's also worth noting that if we expect the keychain to always be available, there may be a keychain issue here.

Reply:
@OrlaM @nbhasin2 Ya'll have any idea re: this before I go investigating?

Reply:
With the work week last week this got lost in my emails, sorry. I'm not sure I understand the distinction between unavailable and missing. Are you asking if the key chain can possibly be in a state where it potentially has data but is unresponsive?

Reply:
No worries! I'm wondering if it is correct to assume that if a call to retrieve a keychain value returns nil, all subsequent calls will return nil unless that keychain value is reset. Right now we are assuming that once a call to `MZKeychainWrapper.sharedClientAppContainerKeychain.string` for the logins encryption key returns nil it will always return nil unless we reset the key value via `MZKeychainWrapper.sharedClientAppContainerKeychain.set`.

Subsequently, we wipe the database because any saved logins can't be decrypted without the old key. However, if a keychain value can be temporarily unavailable, wiping the database would be heavy-handed and result in unnecessary data loss for the user.

Reply:
So, from my investigation yesterday, there are cases where the key-chain can be unavailable (but still contain a value). However, everything I've uncovered so far seems to indicate that those are edge cases, are pretty rare, and, if you're running into them, there's something else that's generally wrong, the the assumption seems to be, from what I'm reading, that keychain is generally available, and a value that returns `nil` will return `nil` until it's set.

However, I'll let people who'eve worked with keychain answer more clearly. I've literally only looked into this for about an hour yesterday.

Reply:
> No worries! I'm wondering if it is correct to assume that if a call to retrieve a keychain value returns nil, all subsequent calls will return nil unless that keychain value is reset. 
This is not a safe assumption. There is at least one known issue with the keychain where it can return nil for data that is present, from a little googling it seems like this issue is still present today. When the app is resumed from a background state it can take a little time for the OS to decrypt any protected data and if a call to fetch that data is made before the decryption process has finished it will return nil, but trying again at a later time will return the expected value. 

A common work around is to use this value to check first if the data is available, and linked on the same page are notifications you can observe to find out when the data becomes available. 
https://developer.apple.com/documentation/uikit/uiapplication/1622925-isprotecteddataavailable

Reply:
Since `getStoredKey` is called when users attempt to add, update, or sync logins I think the best solution at present is to add a guard clause for`isProtectedDataAvailable` and if it fails return an error that the caller can process as needed. 

For syncing this will mean logins are excluded from syncs until the keychain is available which, because of how often syncs are triggered, should be imperceptible to the user as long as the keychain is eventually available. 
But for adding and updating logins, it will mean that while the keychain is unavailable those calls will fail as there's no way to perform the encryption without the key. And the failure could be perceived as data loss.

I'm hesitant to add observers here as I'm not sure how these database calls being triggered after a notification is sent will play with the concurrent code in place. I would definitely be concerned about unintended side effects. Let me know if this makes sense or if you have better suggestions. @OrlaM @adudenamedruby 

And thanks for looking into this!

Reply:
Makes sense to me 👍

Reply:
@adudenamedruby After meeting with @OrlaM, I removed the `isProtectedDataAvailable` check as it's difficult to retrieve this value for use in the backend. We can reassess this decision if needed. But for now we should be able to remove the do not merge label and attempt another bitrise build.